src/main/java/cn/nukkit/blockentity/BlockEntityPistonArm.java

```diff
@@ -40,45 +49,154 @@ protected void initBlockEntity() {
 
-        if (namedTag.contains("Extending")) {
-            this.extending = namedTag.getBoolean("Extending");
-        }
 
-        if (namedTag.contains("powered")) {
-            this.powered = namedTag.getBoolean("powered");
+        if (namedTag.contains("facing")) {
+            this.facing = BlockFace.fromIndex(namedTag.getInt("facing"));
+        } else {
+            Block b = this.getLevelBlock();
+
+            if (b instanceof Faceable) {
+                this.facing = ((Faceable) b).getBlockFace();
+            } else {
+                this.facing = BlockFace.NORTH;
+            }
         }
 
+        attachedBlocks = new ArrayList<>();
+
         if (namedTag.contains("AttachedBlocks")) {
             ListTag blocks = namedTag.getList("AttachedBlocks", IntTag.class);
             if (blocks != null && blocks.size() > 0) {
-                this.attachedBlock = new Vector3((double) ((IntTag) blocks.get(0)).getData(), (double) ((IntTag) blocks.get(1)).getData(), (double) ((IntTag) blocks.get(2)).getData());
+                for (int i = 0; i < blocks.size(); i += 3) {
+                    this.attachedBlocks.add(new BlockVector3(
+                            ((IntTag) blocks.get(i)).data,
+                            ((IntTag) blocks.get(i + 1)).data,
+                            ((IntTag) blocks.get(i + 1)).data
+                    ));
+                }
             }
         } else {
-            namedTag.putList(new ListTag("AttachedBlocks"));
+            namedTag.putList(new ListTag<>("AttachedBlocks"));
         }
 
         super.initBlockEntity();
     }
 
-    private void pushEntities() {
-        float lastProgress = this.getExtendedProgress(this.lastProgress);
-        double x = (double) (lastProgress * (float) this.facing.getXOffset());
-        double y = (double) (lastProgress * (float) this.facing.getYOffset());
-        double z = (double) (lastProgress * (float) this.facing.getZOffset());
-        AxisAlignedBB bb = new SimpleAxisAlignedBB(x, y, z, x + 1.0D, y + 1.0D, z + 1.0D);
+    private void moveCollidedEntities() {
+        BlockFace pushDir = this.extending ? facing : facing.getOpposite();
+        for (BlockVector3 pos : this.attachedBlocks) {
+            BlockEntity blockEntity = this.level.getBlockEntity(pos.getSide(pushDir));
+
+            if (blockEntity instanceof BlockEntityMovingBlock) {
+                ((BlockEntityMovingBlock) blockEntity).moveCollidedEntities(this, pushDir);
+            }
+        }
+
+        AxisAlignedBB bb = new SimpleAxisAlignedBB(0, 0, 0, 1, 1, 1).getOffsetBoundingBox(
+                this.x + (pushDir.getXOffset() * progress),
+                this.y + (pushDir.getYOffset() * progress),
+                this.z + (pushDir.getZOffset() * progress)
+        );
+
         Entity[] entities = this.level.getCollidingEntities(bb);
-        if (entities.length != 0) {
 
+        for (Entity entity : entities) {
+            moveEntity(entity, pushDir);
```

Comment:
The painting should drop when it is pushed by the piston

Reply:
Thanks, fixed